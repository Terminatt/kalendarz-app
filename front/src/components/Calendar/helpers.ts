import { SwitcherOption } from '@components/Switcher/Switcher';
import { DAY_NAMES, Month } from '@constants/constants';
import {
    isBeforeToday, isToday, isWeekend,
} from '@utils/general';
import dayjs, { Dayjs } from 'dayjs';
import { CalendarItemType } from './CalendarItem/CalendarItem';

export interface CalendarDay {
    dayNumber: string;
    dayName: string;
    type: CalendarItemType;
    date: Dayjs;
}
export type EvaluateTypeHandler = (date: Dayjs) => CalendarItemType;
export function createDayList(year: number, month: Month, evaluateType: EvaluateTypeHandler): CalendarDay[] {
    const dayList: CalendarDay[] = [];
    const date = new Date(year, month, 1);
    const numberOfDays = dayjs(date).daysInMonth();

    for (let i = 0; i < numberOfDays; i++) {
        const dateTmp = dayjs(date).add(i, 'day');
        const dayNumber = dateTmp.day();
        dayList.push(
            {
                dayNumber: i < 9 ? `0${i + 1}` : `${i + 1}`,
                dayName: DAY_NAMES[dayNumber],
                type: evaluateType(dateTmp),
                date: dateTmp,
            },
        );
    }
    return dayList;
}

export function createLastWeekList(year: number, month: Month, evaluateType: EvaluateTypeHandler): CalendarDay[] {
    const dayList: CalendarDay[] = [];
    const date = new Date(year, month, 1);
    const lastDay = dayjs(date).endOf('month');
    const lastDayNumber = lastDay.day();

    for (let i = lastDayNumber - 1; i >= 0; i--) {
        const dateTmp = dayjs(lastDay).subtract(i, 'day');
        const dayNumber = dateTmp.day();
        dayList.push(
            {
                dayNumber: `${dateTmp.format('DD')}`,
                dayName: DAY_NAMES[dayNumber],
                type: evaluateType(dateTmp),
                date: dateTmp,
            },
        );
    }
    return dayList;
}

export function evaluateCurrentMonthDayType(current: Dayjs): CalendarItemType {
    if (isBeforeToday(current)) {
        return CalendarItemType.BEFORE_TODAY;
    }

    if (isToday(current)) {
        return CalendarItemType.TODAY;
    }

    if (isWeekend(current)) {
        return CalendarItemType.DAYOFF;
    }

    return CalendarItemType.NORMAL;
}

export function getMonthsOptions(months: string[]): SwitcherOption<Month>[] {
    return months.map((el, index) => ({ label: el, value: index }));
}
