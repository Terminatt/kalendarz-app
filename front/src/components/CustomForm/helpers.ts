import { ResponseError } from '@generics/generics';
import { getEntries } from '@utils/general';

export function parseErrorResponse(errorResponse: ResponseError, ERROR_MESSAGES: {[key: string]: string}):
Array<{name: string, errors: string[]}> {
    const entries = getEntries(errorResponse);
    const errors: Array<{name: string, errors: string[]}> = [];

    // TODO SERVE ARRAY CASE
    entries.forEach((el) => {
        if (el[1] && !Array.isArray(el[1])) {
            errors.push({ name: el[0], errors: [ERROR_MESSAGES[el[1].type]] });
        }
    });

    return errors;
}
