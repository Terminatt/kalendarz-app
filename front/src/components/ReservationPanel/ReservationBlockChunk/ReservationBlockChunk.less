@import '~styles/variables.less';

.block {
    background-color: @reservation-block-color;
    cursor: pointer;
    border-radius: @border-radius-regular;
    transition: background-color @animation-duration-base;
    position: relative;

    &-btn {
        background-color: inherit;
        border: none;
        width: 100%;
        cursor: pointer;


        &-text {
            visibility: hidden;
            cursor: pointer;
        }

        &:focus {
            outline: none;
        }
    }

    &-selected {
        background-color: @primary-color;
    }

    &:hover, &:focus {
        background-color: @primary-color;

        .block-overlay {
            display: block;
            opacity: 1;
        }
    }

    &-overlay {
        width: max-content;
        position: absolute;
        background-color: @secondary-bg;
        padding: @padding-tiny;
        border-radius: @border-radius-regular;
        display: none;
        opacity: 0;
        transition: opacity @animation-duration-base;
        top: -50px;
        z-index: 100; 
        left: 50%;
        transform: translateX(-50%);

        &-arrow {
            background-color: @secondary-bg;
            position: absolute;
            bottom: -10px;
            left: 50%;
            transform: translateX(-50%);
            width: 50px;
            height: 30px;
            clip-path: polygon(49% 100%, 0 0, 100% 0);
            z-index: -1; 
        }
    }
}