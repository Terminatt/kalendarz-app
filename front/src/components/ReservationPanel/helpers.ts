import { TIME_BLOCK_MINUTES } from '@constants/constants';
import { HoveredFocusedTimeBlock, TimeInterval } from './ReservationPanel';

export function getTimeBlockValue(timeBlockIndex: number): string {
    const timeValue = TIME_BLOCK_MINUTES * timeBlockIndex;
    return timeValue === 0 ? '00' : timeValue.toString();
}

export function isTimeBlockSelected(timeValue: number, interval?: TimeInterval, hoveredFocused?: HoveredFocusedTimeBlock): boolean {
    if (!interval?.start) {
        return false;
    }

    if (hoveredFocused?.hovered) {
        return timeValue <= hoveredFocused.hovered;
    }

    if (hoveredFocused?.focused) {
        return timeValue <= hoveredFocused.focused;
    }

    if (interval?.end) {
        return timeValue >= interval.start && timeValue <= interval.end;
    }

    return timeValue === interval.start;
}
