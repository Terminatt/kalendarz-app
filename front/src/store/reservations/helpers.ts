import { BLOCK_COLORS } from '@constants/constants';
import { getRandomItem } from '@utils/general';
import dayjs from 'dayjs';
import { Reservation, ReservationHashMap } from './types';

export function parseReservationData(reservations: Reservation[]): ReservationHashMap {
    const hashMapData: ReservationHashMap = {};
    reservations.forEach((el) => {
        if (!hashMapData[el.room]) {
            hashMapData[el.room] = [];
        }

        hashMapData[el.room].push({
            ...el,
            start: dayjs(el.start),
            end: dayjs(el.end),
            color: getRandomItem(BLOCK_COLORS),
        });
    });

    return hashMapData;
}
