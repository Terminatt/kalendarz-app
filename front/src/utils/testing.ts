export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
