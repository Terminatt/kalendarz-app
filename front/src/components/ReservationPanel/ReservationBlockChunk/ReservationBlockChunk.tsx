/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import { TIME_BLOCK_MINUTES } from '@constants/constants';
import { joinClassNames, parseHourDate } from '@utils/general';
import dayjs, { Dayjs } from 'dayjs';
import React, { useEffect, useState } from 'react';
import { TimeIntervalWithRoom } from '../ReservationPanel';
import { isBlockSelected } from './helpers';

import './ReservationBlockChunk.less';

export interface ReservationBlockChunkProps {
    start: Dayjs;
    blocks: number;
    selectedInterval?: TimeIntervalWithRoom;
    hoveredEnd?: Dayjs;
    isPast?: boolean;
    onClick?: (startLimit: Dayjs, endLimit: Dayjs, selected: Dayjs) => void;
    onMouseEnter?: (startLimit: Dayjs, endLimit: Dayjs, selected: Dayjs) => void;
    onMouseLeave?: () => void;
}

const ReservationBlockChunk: React.FC<ReservationBlockChunkProps> = (props) => {
    const [end, setEnd] = useState<Dayjs>(dayjs());
    const {
        start, blocks, selectedInterval, hoveredEnd, isPast, onClick, onMouseEnter, onMouseLeave,
    } = props;

    useEffect(() => {
        setEnd(start.minute(TIME_BLOCK_MINUTES * blocks));
    }, [start, blocks]);

    const createBlocks = () => {
        const elements: React.ReactElement[] = [];
        let time = start;
        for (let i = 0; i <= blocks; i++) {
            const displayTime = parseHourDate(time);
            const selected = isBlockSelected(time, selectedInterval, hoveredEnd);
            const cloned = time.clone();
            elements.push(
                <td
                    data-testid={`block-${i}`}
                    onMouseEnter={() => !isPast && onMouseEnter && onMouseEnter(start, end, cloned)}
                    onMouseLeave={() => !isPast && onMouseLeave && onMouseLeave()}
                    onClick={() => !isPast && onClick && onClick(start, end, cloned)}
                    key={displayTime}
                    colSpan={2}
                    className={joinClassNames(['block', selected ? 'block-selected' : null, isPast ? 'block-disabled' : null])}
                >
                    <button
                        data-testid={`btn-${i}`}
                        disabled={isPast}
                        onFocus={() => !isPast && onMouseEnter && onMouseEnter(start, end, cloned)}
                        onBlur={() => !isPast && onMouseLeave && onMouseLeave()}
                        type="button"
                        className="block-btn"
                    >
                        <span className="block-btn-text">
                            {displayTime}
                        </span>
                    </button>
                    <div className="block-overlay">
                        {displayTime}
                        <div className="block-overlay-arrow" />
                    </div>
                </td>,
            );
            const minute = time.get('minute');
            time = time.minute(minute + TIME_BLOCK_MINUTES);
        }
        return elements;
    };

    return (
        <>
            {createBlocks()}
        </>
    );
};

export default ReservationBlockChunk;
