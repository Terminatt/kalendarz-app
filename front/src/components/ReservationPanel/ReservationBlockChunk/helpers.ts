import { Dayjs } from 'dayjs';
import { TimeIntervalWithRoom } from '../ReservationPanel';

export function isBlockSelected(current: Dayjs, selectedInterval?: TimeIntervalWithRoom, hovered?: Dayjs): boolean {
    if (current.isSame(hovered)) {
        return true;
    }

    if (selectedInterval?.start?.isBefore(current) && selectedInterval?.end?.isAfter(current)) {
        return true;
    }

    if (selectedInterval?.start?.isSame(current) || selectedInterval?.end?.isSame(current)) {
        return true;
    }

    if (!selectedInterval?.start || !hovered) {
        return false;
    }

    if (selectedInterval.startLimit.isAfter(hovered) || selectedInterval.endLimit.isBefore(hovered)) {
        return false;
    }

    if (current.isSameOrAfter(hovered) && current.isBefore(selectedInterval.start)) {
        return true;
    }

    if (!selectedInterval.end) {
        if (current.isSameOrBefore(hovered) && current.isAfter(selectedInterval.start)) {
            return true;
        }
    } else if (current.isSameOrBefore(hovered) && current.isAfter(selectedInterval.end)) {
        return true;
    }

    return false;
}
