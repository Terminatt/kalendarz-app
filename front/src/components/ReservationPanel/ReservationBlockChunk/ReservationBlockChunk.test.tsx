import React from 'react';
import renderer from 'react-test-renderer';
import { Menu } from 'antd';
import ReservationBlockChunk from './ReservationBlockChunk';
import MockDate from 'mockdate';
import dayjs from 'dayjs';
import { fireEvent, render } from '@utils/testing';
import { isBlockSelected } from './helpers';

const date = '2022-02-20';
describe('Reservation Block Chunk Component', () => {
    
    beforeEach(() => {
        MockDate.set(date);
    });

    afterEach(() => {
        MockDate.reset();
    });

    it('matches the snapshot', () => {
        const tree = renderer
            .create(
                <table>
                    <tbody>
                        <tr>
                            <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} />
                        </tr>
                    </tbody>
                </table>
            )
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('should trigger on click when clicking on block', async () => {
        const onClick= jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} onClick={onClick} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('block-1');
        fireEvent.click(btn)
        expect(onClick).toBeCalledTimes(1);
    });

    it('should not trigger on click when clicking on block', async () => {
        const onClick= jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} isPast onClick={onClick} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('block-1');
        fireEvent.click(btn)
        expect(onClick).not.toBeCalled();
    });

    it('should trigger on click when clicking on block button', async () => {
        const onClick= jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} onClick={onClick} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('btn-1');
        fireEvent.click(btn)
        expect(onClick).toBeCalledTimes(1);
    });

    it('should not trigger on click when clicking on block button', async () => {
        const onClick= jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} isPast onClick={onClick} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('btn-1');
        fireEvent.click(btn)
        expect(onClick).not.toBeCalled();
    });

    it('should trigger on mouse enter', async () => {
        const onMouseEnter = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} onMouseEnter={onMouseEnter} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('block-1');
        fireEvent.mouseEnter(btn)
        expect(onMouseEnter).toBeCalledTimes(1);
    });

    it('should trigger on mouse enter on button', async () => {
        const onMouseEnter = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} onMouseEnter={onMouseEnter} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('btn-1');
        fireEvent.focus(btn)
        expect(onMouseEnter).toBeCalledTimes(1);
    });

    it('should not trigger on mouse enter', async () => {
        const onMouseEnter = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} isPast onMouseEnter={onMouseEnter} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('block-1');
        fireEvent.mouseEnter(btn)
        expect(onMouseEnter).not.toBeCalled();
    });

    it('should not trigger on mouse enter on button', async () => {
        const onMouseEnter = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} isPast onMouseEnter={onMouseEnter} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('btn-1');
        fireEvent.focus(btn)
        expect(onMouseEnter).not.toBeCalled();
    });

    it('should trigger on mouse leave', async () => {
        const onMouseLeave = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} onMouseEnter={onMouseLeave} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('block-1');
        fireEvent.mouseEnter(btn)
        expect(onMouseLeave).toBeCalledTimes(1);
    });

    it('should trigger on mouse leave on button', async () => {
        const onMouseLeave = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} onMouseEnter={onMouseLeave} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('btn-1');
        fireEvent.focus(btn)
        fireEvent.blur(btn)
        expect(onMouseLeave).toBeCalledTimes(1);
    });

    it('should not trigger on mouse leave', async () => {
        const onMouseLeave = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} isPast onMouseEnter={onMouseLeave} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('block-1');
        fireEvent.mouseEnter(btn)
        expect(onMouseLeave).not.toBeCalled();
    });

    it('should not trigger on mouse leave on button', async () => {
        const onMouseLeave = jest.fn();
        const { element } = render(
            <table>
                <tbody>
                    <tr>
                        <ReservationBlockChunk start={dayjs().hour(8)} blocks={4} isPast onMouseEnter={onMouseLeave} />
                    </tr>
                </tbody>
            </table>

        );
        
        const btn = element.getByTestId('btn-1');
    
        fireEvent.focus(btn)
        fireEvent.blur(btn)
        expect(onMouseLeave).not.toBeCalled();
    });
});

describe('Reservation Block Chunk helpers', () => {
    describe('isBlockSelected', () => {
        it('should return true when hovered', () => {
            const selected = isBlockSelected(dayjs(), undefined, dayjs())

            expect(selected).toBe(true);
        });

        it('should return true when is between interval', () => {
            const selected = isBlockSelected(dayjs().hour(2), {
                start: dayjs().hour(1), 
                end: dayjs().hour(3),
                startLimit: dayjs().hour(0),
                endLimit: dayjs().hour(4),
            })

            expect(selected).toBe(true);
        });

        it('should return true when is the same as start', () => {
            const selected = isBlockSelected(dayjs().hour(1), {
                start: dayjs().hour(1), 
                end: dayjs().hour(3),
                startLimit: dayjs().hour(0),
                endLimit: dayjs().hour(4),
            })

            expect(selected).toBe(true);
        });

        it('should return false when is there is no selected interval and hovered', () => {
            const selected = isBlockSelected(dayjs())

            expect(selected).toBe(false);
        });

        it('should return false when hovered is before startLimit', () => {
            const selected = isBlockSelected(dayjs().hour(5), {
                start: dayjs().hour(11), 
                end: dayjs().hour(13),
                startLimit: dayjs().hour(8),
                endLimit: dayjs().hour(15),
            }, dayjs().hour(6))

            expect(selected).toBe(false);
        });

        it('should return false when hovered is after endLimit', () => {
            const selected = isBlockSelected(dayjs().hour(5), {
                start: dayjs().hour(11), 
                end: dayjs().hour(13),
                startLimit: dayjs().hour(8),
                endLimit: dayjs().hour(15),
            }, dayjs().hour(16))

            expect(selected).toBe(false);
        });

        it('should return true when current is between start and hovered', () => {
            const selected = isBlockSelected(dayjs().hour(13), {
                start: dayjs().hour(11), 
                end: dayjs().hour(12),
                startLimit: dayjs().hour(8),
                endLimit: dayjs().hour(15),
            }, dayjs().hour(14))

            expect(selected).toBe(true);
        });

        it('should return true when current is between hovered and start and there is no end', () => {
            const selected = isBlockSelected(dayjs().hour(10), {
                start: dayjs().hour(11), 
                startLimit: dayjs().hour(7),
                endLimit: dayjs().hour(15),
            }, dayjs().hour(8))

            expect(selected).toBe(true);
        });

        it('should return true when current is between hovered and start and there is no end', () => {
            const selected = isBlockSelected(dayjs().hour(10), {
                start: dayjs().hour(11), 
                startLimit: dayjs().hour(7),
                endLimit: dayjs().hour(15),
            }, dayjs().hour(8))

            expect(selected).toBe(true);
        });

        it('should return true when current is between hovered and end', () => {
            const selected = isBlockSelected(dayjs().hour(13), {
                start: dayjs().hour(11), 
                end: dayjs().hour(12),
                startLimit: dayjs().hour(7),
                endLimit: dayjs().hour(15),
            }, dayjs().hour(14))

            expect(selected).toBe(true);
        });

        it('should return false when current is between hovered and start', () => {
            const selected = isBlockSelected(dayjs().hour(10), {
                start: dayjs().hour(11), 
                end: dayjs().hour(12),
                startLimit: dayjs().hour(7),
                endLimit: dayjs().hour(15),
            }, dayjs().hour(9))

            expect(selected).toBe(true);
        });
    })
})
