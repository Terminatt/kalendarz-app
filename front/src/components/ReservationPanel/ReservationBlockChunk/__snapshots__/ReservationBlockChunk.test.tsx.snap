// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reservation Block Chunk Component matches the snapshot 1`] = `
<table>
  <tbody>
    <tr>
      <td
        className="block"
        colSpan={2}
        data-testid="block-0"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="block-btn"
          data-testid="btn-0"
          onBlur={[Function]}
          onFocus={[Function]}
          type="button"
        >
          <span
            className="block-btn-text"
          >
            08:00
          </span>
        </button>
        <div
          className="block-overlay"
        >
          08:00
          <div
            className="block-overlay-arrow"
          />
        </div>
      </td>
      <td
        className="block"
        colSpan={2}
        data-testid="block-1"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="block-btn"
          data-testid="btn-1"
          onBlur={[Function]}
          onFocus={[Function]}
          type="button"
        >
          <span
            className="block-btn-text"
          >
            08:15
          </span>
        </button>
        <div
          className="block-overlay"
        >
          08:15
          <div
            className="block-overlay-arrow"
          />
        </div>
      </td>
      <td
        className="block"
        colSpan={2}
        data-testid="block-2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="block-btn"
          data-testid="btn-2"
          onBlur={[Function]}
          onFocus={[Function]}
          type="button"
        >
          <span
            className="block-btn-text"
          >
            08:30
          </span>
        </button>
        <div
          className="block-overlay"
        >
          08:30
          <div
            className="block-overlay-arrow"
          />
        </div>
      </td>
      <td
        className="block"
        colSpan={2}
        data-testid="block-3"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="block-btn"
          data-testid="btn-3"
          onBlur={[Function]}
          onFocus={[Function]}
          type="button"
        >
          <span
            className="block-btn-text"
          >
            08:45
          </span>
        </button>
        <div
          className="block-overlay"
        >
          08:45
          <div
            className="block-overlay-arrow"
          />
        </div>
      </td>
      <td
        className="block"
        colSpan={2}
        data-testid="block-4"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="block-btn"
          data-testid="btn-4"
          onBlur={[Function]}
          onFocus={[Function]}
          type="button"
        >
          <span
            className="block-btn-text"
          >
            09:00
          </span>
        </button>
        <div
          className="block-overlay"
        >
          09:00
          <div
            className="block-overlay-arrow"
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
